cnd/src/db/mod.rs

```diff
@@ -0,0 +1,173 @@
+mod models;
+mod schema;
+
+pub use crate::db::models::Metadata;
+
+use crate::{
+    db::models::NewMetadata,
+    swap_protocols::{
+        metadata_store::{self, AssetKind, Error, LedgerKind, MetadataStore, Role},
+        SwapId,
+    },
+};
+use diesel::{self, prelude::*, sqlite::SqliteConnection};
+use libp2p::PeerId;
+use migrations_internals;
+use std::{
+    convert::TryFrom,
+    fs::File,
+    io::stdout,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+/// This module provides an Sqlite backed persistent MatadataStore.
+
+#[derive(Debug)]
+pub struct SqliteMetadataStore {
+    db: PathBuf,
+}
+
+impl SqliteMetadataStore {
+    /// Return a handle that can be used to access the database.
+    ///
+    /// When this returns an Sqlite database exists at 'db' and connection to
+    /// the database has been verified.
+    pub fn new(db: Option<PathBuf>) -> Result<SqliteMetadataStore, Error> {
+        let db = db
+            .or(default_db_path())
+            .ok_or_else(|| Error::Path("Failed to get database path".to_string()))?;
+
+        create_database_if_needed(&db)?;
+        log::info!("Sqlite database file: {}", db.display());
+
+        let migrations = default_migrations_dir()
+            .ok_or_else(|| Error::Path("Failed to get migrations directory".to_string()))?;
+
+        create_migrations_directory_if_needed(&migrations)?;
+        run_migrations(&db, &migrations)?;
+
+        Ok(SqliteMetadataStore {
+            db: db.clone().to_path_buf(),
+        })
+    }
+}
+
+impl MetadataStore for SqliteMetadataStore {
+    fn insert<M: Into<metadata_store::Metadata>>(
```

Comment:
Any particular reason why this needs to be `<M: Into<...>` and not just `Metadata`?

Reply:
That was in the code before I touched it.  I'm not sure what is the full benefit of having it or why we added the trait instead of a concrete type.

Reply:
I removed it 😁